### PR TITLE
http_client: use slist function for NO_PROXY parsing

### PR DESCRIPTION
**Pull request by erain**
This is a follow up PR for https://github.com/fluent/fluent-bit/pull/3272
- make use the functions from `slist`
- enhance the unit tests for `slist`

**Documentation**

- [x] Documentation required for this feature

https://github.com/fluent/fluent-bit-docs/pull/532

### CODE REVIEW

File: src/flb_config.c [hunk 1]
```diff
@@ -46,6 +46,16 @@ const char *FLB_CONF_ENV_LOGLEVEL = "FLB_LOG_LEVEL";
...
 
+/* emptyval checks whether a string has a non-null value "". */
+static inline int flb_str_emptyval(const char *s)
+{
+    if (s != NULL && strcmp(s, "") == 0) {
+        return FLB_TRUE;
+    }
+    return FLB_FALSE;
+}
+
```
Comment (JeffLuoo) [inline src/flb_config.c:57]: extra newline?

**Comment by erain** [comment]
@edsiper @qingling128 I just found that I missed pushing a commit (different dev machine etc.) for the original PR. 

Sorry for the mess up. Could you help review this PR and cherry-pick it for 1.7.5 release if possible? 

Thanks!!

File: src/flb_config.c (hunk 1 above)
Comment (erain) [reply src/flb_config.c:57]: Done

**Comment by erain** [comment]
I've also added the documentation PR in https://github.com/fluent/fluent-bit-docs/pull/532

**Comment by ekund** [comment]
I have tested this as well, and it addresses the issues I've been seeing, in https://github.com/fluent/fluent-bit/issues/3483.

**Comment by qingling128** [comment]
@erain Unfortunately 1.7.5 is out already. We are discussing offline the potential to do a 1.7.6 release next week for this.

**Comment by erain** [comment]
Thanks @qingling128 and releasing it in 1.7.6 sounds great.

**Review by qingling128 — DISMISSED** [review]
LGTM

**Comment by erain** [comment]
@edsiper could you help review this PR as well as the doc PR in https://github.com/fluent/fluent-bit-docs/pull/532 ?

File: src/flb_config.c [hunk 2]
```diff
@@ -46,6 +46,15 @@ const char *FLB_CONF_ENV_LOGLEVEL = "FLB_LOG_LEVEL";
 
 int flb_regex_init();
 
+/* emptyval checks whether a string has a non-null value "". */
```
Comment (edsiper) [inline src/flb_config.c:49]: why the string function was moved to flb_config.c ?

File: src/flb_upstream.c [hunk 3]
```diff
@@ -258,15 +264,29 @@ int flb_should_proxy_for_host(const char *host, const char *proxy, const char *n
...
+         if (strcmp(host, e->str)) {
+            ret = FLB_FALSE;
+            break;
         }
-        no_proxy_url = strtok(NULL, ",");
     }
 
-    return FLB_TRUE;
+    /* clean up the resources. */
+    if (&no_proxy_list) {
```
Comment (edsiper) [inline src/flb_upstream.c:285]: this check is invalid, ```no_proxy_list``` is a variable from the stack, if you want to check if it's empty can use mk_list_size(&no_proxy_list)  == 0 (to check if it has zero nodes linked)

File: src/flb_upstream.c [hunk 4]
```diff
@@ -258,15 +264,29 @@ int flb_should_proxy_for_host(const char *host, const char *proxy, const char *n
...
+        flb_slist_destroy(&no_proxy_list);
```
Comment (edsiper) [inline src/flb_upstream.c:286]: it's safe to destroy an initialized empty list... so the above conditional is not required

**Review by edsiper — CHANGES_REQUESTED** [review]

**Comment by qingling128** [comment]
Thanks @erain ! 

@edsiper - Could you take another look?

**Review by edsiper — APPROVED** [review]

**Pull request merged**